### PR TITLE
Add Number Normalisation for SpeechT5

### LINKED ISSUE #23480: SpeechT5 cannot read numbers
### System Info

transformers == 4.29.0
environment = Colab
Python == 3.10.11
tensorflow == 2.12.0
torch == 2.0.1+cu118
torchaudio == 2.0.2+cu118

### Who can help?

@sanchit-gandhi 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

1. Init a Transformer agent
2. Init a text which contains numbers. For example text = "More than 10 people have been killed by Covid."
3. Call the agent for a text-to-speech (SpeechT5). For example, audio_translated = agent.run("Read out loud the text", text=text)
4. Play the generated audio

The audio blanks all the numbers/digits.

I am suspecting SpeechT5 to behave wrongly as the code generated by the agent seems to be correct.

Good luck :)

### Expected behavior

The audio file should contain numbers/digits indicated in the text.

### PR DESCRIPTION
**Pull request by tanaymeh**
# What does this PR do?
This PR will add number normalisation for SpeechT5. Currently, SpeechT5 cannot read numbers, as described in #23480.

Fixes #23480

# Cases covered so far
- [x] Integer number normalisation
- [x] Floating point number normalisation
- [x] Currency normalisation
	- [x] US Dollar `$`
	- [x] cents `¢`
	- [x] Euro `€`
	- [x] Sterling Pound `£`
	- [x] Yen `¥`
	- [x] Other major currency symbols
- [x] Percents `%` 
- [x] Negative numbers (starting with `-`)
- [x] Numbers from 0 to Decillion
- [x] Any combination of those stated and ticked above

## Who can review?
@sanchit-gandhi

### CODE REVIEW

**Comment by sgugger** [comment]
cc @sanchit-gandhi

File: src/transformers/models/speecht5/number_normalizer.py
Comment (sanchit-gandhi) [inline src/transformers/models/speecht5/number_normalizer.py:20]: This is looking super comprehensive already! Thanks for listing all the normalisations applied in the PR comment - are there any big ones that we're missing from Whisper?

File: src/transformers/models/speecht5/number_normalizer.py
Comment (sanchit-gandhi) [inline src/transformers/models/speecht5/number_normalizer.py:142]: No need for any print statements! Might have been left in from debugging

File: src/transformers/models/speecht5/number_normalizer.py
Comment (sanchit-gandhi) [inline src/transformers/models/speecht5/number_normalizer.py:144]: likewise this comment

File: src/transformers/models/speecht5/number_normalizer.py
Comment (sanchit-gandhi) [inline src/transformers/models/speecht5/number_normalizer.py:148]: Can we not use `self.thousands` here?

File: src/transformers/models/speecht5/number_normalizer.py
Comment (sanchit-gandhi) [inline src/transformers/models/speecht5/number_normalizer.py:93]: Should these have the extra space before?

**Review by sanchit-gandhi — COMMENTED** [review]
Looks very comprehensive already @tanaymeh! Wonder if we could add a few tests to [`test_tokenization_speech_t5.py`](https://github.com/huggingface/transformers/blob/b42010bb1d3cbf262d27e0a328661885be46dfdb/tests/models/speecht5/test_tokenization_speecht5.py#L32) to confirm we get the expected behaviour? E.g. including tests so that we cover each of the normalisation cases and confirm they work as expected (you can test multiple parts in single tests if you want, e.g. numbers and currency in one go)

File: src/transformers/models/speecht5/number_normalizer.py
Comment (tanaymeh) [reply src/transformers/models/speecht5/number_normalizer.py:20]: Thanks! To my knowledge, all cases are covered from Whisper

File: src/transformers/models/speecht5/number_normalizer.py
Comment (tanaymeh) [reply src/transformers/models/speecht5/number_normalizer.py:85]: Will do!

File: src/transformers/models/speecht5/number_normalizer.py
Comment (tanaymeh) [reply src/transformers/models/speecht5/number_normalizer.py:148]: Yes! I actually wrote this function separately while working on it so I didn't integrate it well with the class. Fixed now!

File: src/transformers/models/speecht5/number_normalizer.py
Comment (tanaymeh) [reply src/transformers/models/speecht5/number_normalizer.py:93]: Yes, it is to ensure that `$12` becomes `twelve dollars` and not `twelvedollars`. Adding a space in the symbol to name mapping is easier and less complicated than doing it in the format string which is returned.

File: src/transformers/models/speecht5/processing_speecht5.py
Comment (tanaymeh) [reply src/transformers/models/speecht5/processing_speecht5.py:21]: That makes sense. Just want to confirm, the right place for `if self.normalize` check should be in the `_tokenize()` function right?

https://github.com/huggingface/transformers/blob/b42010bb1d3cbf262d27e0a328661885be46dfdb/src/transformers/models/speecht5/tokenization_speecht5.py#L143

**Comment by tanaymeh** [comment]
@sanchit-gandhi I added a single test which tests all the covered cases. I also fixed an existing test in the `tests/models/speecht5/test_tokenization_speecht5.py` file which previously was just passing the number '92000' around without normalization but now it would normalize it (see below)

https://github.com/tanaymeh/transformers/blob/afa90898b93c58e712d6d7e3726a66d964282930/tests/models/speecht5/test_tokenization_speecht5.py#L164-L183

File: src/transformers/models/speecht5/number_normalizer.py
Comment (sanchit-gandhi) [reply src/transformers/models/speecht5/number_normalizer.py]: Cool cool - thanks for clarifying! Good with me to do it this way

File: tests/models/speecht5/test_tokenization_speecht5.py
Comment (sanchit-gandhi) [inline tests/models/speecht5/test_tokenization_speecht5.py:55]: Nice - thanks for checking all cases!

File: tests/models/speecht5/test_tokenization_speecht5.py
Comment (sanchit-gandhi) [inline tests/models/speecht5/test_tokenization_speecht5.py:69]: Could we make this a bit more clear by setting `normalize=True` for the one we want to normalize?

Is the test also maybe a bit better if we check that the output of the tokenizer is equal to the expected text? E.g.
```python
input_text, expected_text = self.get_numeric_input_output_texts()
input_ids = tokenizer.encode(input_text, normalize=True)
output_text = tokenizer.decode(input_ids, skip_special_tokens=True)
self.assertListEqual(output_text, tokenizer.encode)
```

File: tests/models/speecht5/test_tokenization_speecht5.py
Comment (sanchit-gandhi) [inline tests/models/speecht5/test_tokenization_speecht5.py:168]: Thanks for updating these tests

**Review by sanchit-gandhi — APPROVED** [review]
Thanks for adding the tests! LGTM - would just request that we change the default behaviour to not normalise, and maybe add a line in the SpeechT5 tokenizer docs to highlight to users that they can use this great feature?

File: src/transformers/models/speecht5/tokenization_speecht5.py
Comment (tanaymeh) [reply src/transformers/models/speecht5/tokenization_speecht5.py:103]: That makes sense, having it on `normalize=True` may break existing code where this is used. I have now set it to False.

**Comment by tanaymeh** [comment]
I have a rather dumb doubt, @sanchit-gandhi: I had added the information about the `normalize` argument in the docstring of the `SpeechT5Tokenizer` class ([here](https://github.com/tanaymeh/transformers/blob/afa90898b93c58e712d6d7e3726a66d964282930/src/transformers/models/speecht5/tokenization_speecht5.py#L68-L69)). Should I add it somewhere else to be shown in the docs? Doesn't the doc builder builds the docstrings and renders them in the docs automatically?

Thanks!

File: src/transformers/models/speecht5/number_normalizer.py
Comment (ArthurZucker) [inline src/transformers/models/speecht5/number_normalizer.py:2]: Is speecht5 not from microsoft?

File: src/transformers/models/speecht5/number_normalizer.py
Comment (ArthurZucker) [inline src/transformers/models/speecht5/number_normalizer.py:52]: The list only include certain currencies (of course!) which short list does it correspond to?

File: src/transformers/models/speecht5/tokenization_speecht5.py
Comment (ArthurZucker) [inline src/transformers/models/speecht5/tokenization_speecht5.py:152]: This should go in the `prepare_inputs_for_tokenization`. This way it could support normalizing on the fly with kwargs, and only normalize once if there are special tokens (instead of normalizing each substrings)

File: src/transformers/models/speecht5/tokenization_speecht5.py
Comment (ArthurZucker) [inline src/transformers/models/speecht5/tokenization_speecht5.py:221]: No sure we need a method for this, re-initializing the normalizer on each tokenization.

File: tests/models/speecht5/test_tokenization_speecht5.py
Comment (ArthurZucker) [reply tests/models/speecht5/test_tokenization_speecht5.py:55]: I am guessing that `"8,000"` is for American normalizer? Or should it also be included?

**Review by ArthurZucker — COMMENTED** [review]
Thanks a lot for working on this! Left a few small nits but otherwise looks good!

File: tests/models/speecht5/test_tokenization_speecht5.py
Comment (tanaymeh) [reply tests/models/speecht5/test_tokenization_speecht5.py:55]: Sorry, I didn't quite understand what you were referring to @ArthurZucker

File: src/transformers/models/speecht5/number_normalizer.py
Comment (tanaymeh) [reply src/transformers/models/speecht5/number_normalizer.py:2]: Yes, Apologies, I copy-pasted the comments at the top from another file and forgot to change it.

File: tests/models/speecht5/test_tokenization_speecht5.py
Comment (ArthurZucker) [reply tests/models/speecht5/test_tokenization_speecht5.py:55]: Just saying that some might write `8000` as `8,000` and wondering if this should also be accounted for?

File: src/transformers/models/speecht5/tokenization_speecht5.py
Comment (tanaymeh) [reply src/transformers/models/speecht5/tokenization_speecht5.py:152]: Makes sense, will do that!

File: src/transformers/models/speecht5/tokenization_speecht5.py
Comment (tanaymeh) [reply src/transformers/models/speecht5/tokenization_speecht5.py:221]: I took the inspiration from Whisper. Instead, should I now initialize the normalizer once in the `__init__()` function and use it directly?

File: src/transformers/models/speecht5/number_normalizer.py
Comment (ArthurZucker) [reply src/transformers/models/speecht5/number_normalizer.py:52]: Not at all! Can you just mention in a comment `# top x most popular currencies according to xxxx ` can just help us know why we went for this list!

File: tests/models/speecht5/test_tokenization_speecht5.py
Comment (tanaymeh) [reply tests/models/speecht5/test_tokenization_speecht5.py:55]: Got it! Adding the below line in the next commit which deals with this issue and normalizes any number with comma separation.

```python
text = re.sub(r"(\d+,\d+)", lambda match: match.group(1).replace(",", ""), text)
```

File: src/transformers/models/speecht5/tokenization_speecht5.py
Comment (tanaymeh) [reply src/transformers/models/speecht5/tokenization_speecht5.py:152]: @ArthurZucker Should I make a new function called `prepare_inputs_for_tokenization`?

Comment (ArthurZucker) [reply src/transformers/models/speecht5/tokenization_speecht5.py:152]: It's actually `prepare_for_tokenization` which should be overloaded!

Comment (tanaymeh) [reply src/transformers/models/speecht5/tokenization_speecht5.py:152]: Gotcha. This is what I am doing currently @ArthurZucker;

```python
# ... rest of the code in Tokenizer class
    @property
    def normalizer(self):
        return self.normalizer

    @normalizer.setter
    def normalizer(self):
        self.normalizer = EnglishNumberNormalizer()
# ... other code

    def prepare_for_tokenization(self, text, is_split_into_words=False, **kwargs):
        normalize = kwargs.pop("normalize", self.normalize)
        if is_split_into_words:
            text = " " + text
        if normalize:
            text = self.normalizer(text)
        return (text, kwargs)
```

Am I on the right track / any suggestions?

**Comment by tanaymeh** [comment]
@ArthurZucker I added your suggested changes ([here](https://github.com/tanaymeh/transformers/blob/682bee8d517398b8b9840808ff3a3bb106f6b9fd/src/transformers/models/speecht5/tokenization_speecht5.py#L126-L132) and [here](https://github.com/tanaymeh/transformers/blob/682bee8d517398b8b9840808ff3a3bb106f6b9fd/src/transformers/models/speecht5/tokenization_speecht5.py#L138-L146)).

I also added support for comma-separated numbers [here](https://github.com/tanaymeh/transformers/blob/682bee8d517398b8b9840808ff3a3bb106f6b9fd/src/transformers/models/speecht5/number_normalizer.py#L185-L186).

Please let me know if I missing something or need to change something.

**Review by ArthurZucker — APPROVED** [review]
Looks good to me! Thanks for iterating 🤗

**Comment by tanaymeh** [comment]
Thanks for your review @sanchit-gandhi and @ArthurZucker. Waiting for review from @ylacombe!

**Comment by ramkrishna757575** [comment]
I am rather new to python and transformers.
Can someone please provide me documentation for this? I am doing the following, but I am not sure where I can pass the normalize parameter.
Currently the audio being generated skips all the numbers given in the input text.

```
import sounddevice as sd
import torch
from datasets import load_dataset
from transformers import SpeechT5Processor, SpeechT5HifiGan, SpeechT5ForTextToSpeech

processor = SpeechT5Processor.from_pretrained("microsoft/speecht5_tts")
embeddings_dataset = load_dataset("Matthijs/cmu-arctic-xvectors", split="validation")
speaker_embeddings = torch.tensor(embeddings_dataset[7306]["xvector"]).unsqueeze(0)
device = "cuda:0" if torch.cuda.is_available() else "cpu"


def synthesise(text):
    inputs = processor(text=text, return_tensors="pt")
    vocoder = SpeechT5HifiGan.from_pretrained("microsoft/speecht5_hifigan").to(device)
    model = SpeechT5ForTextToSpeech.from_pretrained("microsoft/speecht5_tts").to(device)
    speech = model.generate_speech(
        inputs["input_ids"].to(device), speaker_embeddings.to(device), vocoder=vocoder
    )
    return speech.cpu()


def playback_speech(text):
    audio = synthesise(text)
    sd.play(audio.numpy(), samplerate=16000)
    sd.wait()


playback_speech("Numbers from one through ten : 1 2 3 4 5 6 7 8 9 10")
```

**Comment by tanaymeh** [comment]
@ramkrishna757575 You need to turn `normalize=True` in the `processor`. By default, the number normalization is turned off to ensure backwards compatibility since it's a breaking change.

Inside the `synthesise()` function, you can do:
```python
# ... Rest of the code
def synthesise(text):
    inputs = processor(text=text, return_tensors="pt", normalize=True)
# ... Rest of the code
```

**Comment by ramkrishna757575** [comment]
> 

Thanks for the quick reply @tanaymeh.
I tried it, but I got this error:

`Keyword arguments {'normalize': True} not recognized.`

I checked the version of `transformers` that I have installed and it is currently `4.32.0`

**Comment by tanaymeh** [comment]
@ramkrishna757575 About the numbers not being spelt, I am not sure what the reason is. It is certainly not with the processor or tokenizer since both are working fine. See the attached screenshot below.

Screenshot 2023-08-26 at 7 17 14 PM

**Comment by tanaymeh** [comment]
@ramkrishna757575 You need to install the latest version of the transformers (that is not yet released), using:
`pip install git+https://github.com/huggingface/transformers`, in this, you will be able to use this setting.

**Comment by ramkrishna757575** [comment]
ok..thanks @tanaymeh .
I'll try this out.
Will post my findings later once I'm done.

**Comment by sanchit-gandhi** [comment]
Thanks for helping out here @tanaymeh! Here's the parameter in the docs (on `main`) @ramkrishna757575: https://huggingface.co/docs/transformers/main/en/model_doc/speecht5#transformers.SpeechT5Tokenizer.normalize

**Comment by ramkrishna757575** [comment]
> Thanks for helping out here @tanaymeh! Here's the parameter in the docs (on `main`) @ramkrishna757575: https://huggingface.co/docs/transformers/main/en/model_doc/speecht5#transformers.SpeechT5Tokenizer.normalize

got it...thanks @sanchit-gandhi

**Pull request merged**
